pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,607 @@
+		}
+		p.log.Debug("Found aliases", "num_aliases", len(aliasesResp.Aliases))
+
+		for _, alias := range aliasesResp.Aliases {
+			spireKeyID, ok := p.spireKeyIDFromAlias(alias)
+			if !ok {
+				continue
+			}
+			if alias.TargetKeyId == nil {
+				// this means something external to the plugin created the alias, without associating it to a key.
+				// it should never happen with CMKs.
+				return nil, newErrorf(codes.FailedPrecondition, "failed to build cache of KMS keys: found SPIRE alias without key: %q", *alias.AliasName)
+			}
+
+			aliasName := *alias.AliasName
+			g.Go(func() error {
+				entry, err := p.fetchKeyEntry(ctx, aliasName, spireKeyID)
+				if err != nil {
+					return err
+				}
+				keyEntriesMutex.Lock()
+				keyEntries = append(keyEntries, entry)
+				keyEntriesMutex.Unlock()
+				return nil
+			})
+		}
+
+		if !paginator.HasMorePages() {
+			break
+		}
+	}
+
+	if err := g.Wait(); err != nil {
+		statusErr := status.Convert(err)
+		return nil, newErrorf(statusErr.Code(), "failed to build cache of KMS keys: %v", statusErr.Message())
+	}
+
+	return keyEntries, nil
+}
+
+// buildCache gets all aliases from KMS and builds an in memory representation with the ones that belong to the current server
+// for each key that belongs to the server we will trigger a gouroutine that gets the extra information
+func (p *Plugin) buildCache(ctx context.Context) error {
+	keyEntries, err := p.fetchAliasesPages(ctx)
+	if err != nil {
+		return err
+	}
+
+	//add results to cache
+	for _, e := range keyEntries {
+		err := p.setEntry(e.PublicKey.Id, *e)
+		if err != nil {
+			return newError(codes.Internal, err.Error())
+		}
+		p.log.Debug("Key loaded", keyIDTag, e.KMSKeyID, aliasTag, e.Alias)
+	}
+
+	return nil
+}
+
+// scheduleDeleteTask ia a long running task that deletes keys that were rotated
+func (p *Plugin) scheduleDeleteTask(ctx context.Context) {
+	backoffMin := 1 * time.Second
+	backoffMax := 60 * time.Second
+	backoff := backoffMin
+
+	for {
+		select {
+		case <-ctx.Done():
+			p.log.Debug("Stopping schedule delete task", "reason", ctx.Err())
+			p.notifyDelete()
+			return
+		case keyID := <-p.scheduleDelete:
+			log := p.log.With(keyIDTag, keyID)
+			_, err := p.kmsClient.ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
+				KeyId:               aws.String(keyID),
+				PendingWindowInDays: aws.Int32(7),
+			})
+
+			if err == nil {
+				log.Debug("Key deleted")
+				backoff = backoffMin
+				p.notifyDelete()
+				continue
+			}
+
+			var notFoundErr *types.NotFoundException
+			if errors.As(err, &notFoundErr) {
+				log.Error("No such key, dropping from delete schedule")
+				p.notifyDelete()
+				continue
+			}
+
+			var invalidArnErr *types.InvalidArnException
+			if errors.As(err, &invalidArnErr) {
+				log.Error("Invalid ARN, dropping from delete schedule")
+				p.notifyDelete()
+				continue
+			}
+
+			log.Error("It was not possible to schedule key for deletion", "reason", err)
+			p.scheduleDelete <- keyID
```

Comment:
This can lead to deadlock. If the buffered channel gets filled in between the time where the select above receives on this channel and here, then the keyID will block on reinsertion.